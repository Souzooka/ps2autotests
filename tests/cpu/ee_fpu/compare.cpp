#include <stdio.h>
#include <tamtypes.h>

union FloatBits {
	float f;
	u32 u;
};

static u32 CF_ZERO[1] = {0x00000000};
static u32 CF_NEGZERO[1] = {0x80000000};
static u32 CF_MAX[1] = {0x7FFFFFFF};
static u32 CF_MIN[1] = {0xFFFFFFFF};
static u32 CF_MAX_MANTISSA[1] = {0x3FFFFFFF};
static u32 CF_MAX_EXP[1] = {0x7F800001};
static u32 CF_MIN_EXP[1] = {0x00000001};
static u32 CF_NEGONE[1] = {0xBF800000};
static u32 CF_ONE[1] = {0x3F800000};
static u32 CF_GARBAGE1[4] = {0x00001337};
static u32 CF_GARBAGE2[4] = {0xDEADBEEF};

#define FCMP_OP_FUNC(OP) \
static inline void FCMP_##OP(register float &fd, const register float &fs) { \
	asm volatile ( \
		"c." #OP ".s %0, %1\n" \
		: "+f"(fd) : "f"(fs) \
	); \
}

template <u32 i>
static inline void SET_U32(register float &fd) {
	asm volatile (
		"lui $t6, %1\n"
		"ori $t6, $t6, %2\n"
		"mtc1 $t6, %0\n"
		: "+f"(fd) : "K"((i >> 16) & 0xFFFF), "K"(i & 0xFFFF) : "t6"
	);
}

static inline void SET_M(register float &fd, void *p) {
	fd = *(float *)p;
}

static inline void printFloatString(const FloatBits &bits) {
	// We want -0.0 and -NAN to show as negative.
	// So, let's just print the sign manually with an absolute value.
	FloatBits positive = bits;
	positive.u &= ~0x80000000;

	char sign = '+';
	if (bits.u & 0x80000000) {
		sign = '-';
	}

	printf("%c%0.2f", sign, positive.f);
}

static inline void PRINT_F(const register float &ft, bool newline) {
	static FloatBits result = {0};

	asm volatile (
		"swc1 %0, 0(%1)\n"
		: : "f"(ft), "r"((u32)&result.f)
	);

	printf("%08x/", result.u);
	printFloatString(result);
	if (newline) {
		printf("\n");
	}
}

static inline void PRINT_FC(bool newline) {
	register u32 fcr31 = 0;

	asm volatile (
		"cfc1 %0, $31\n"
		: "+r"(fcr31)
	);

	if ((fcr31 >> 23) & 1) {
		printf("true");
	} else {
		printf("false");
	}
	if (newline) {
		printf("\n");
	}
}

#define FCMP_OP_DO_II(OP, s, t) \
	SET_U32<s>(fs); SET_U32<t>(ft); \
	FCMP_##OP(fs, ft); \
	printf("  %s ", #OP); \
	PRINT_F(fs, false); \
	printf(", "); \
	PRINT_F(ft, false); \
	printf(": "); \
	PRINT_FC(true);

#define FCMP_OP_DO_MM(OP, s, t) \
	SET_M(fs, s); SET_M(ft, t); \
	FCMP_##OP(fs, ft); \
	printf("  %s %s, %s: ", #OP, #s, #t); \
	PRINT_FC(true);

#define TEST_FCMP_FUNC(OP) \
FCMP_OP_FUNC(OP); \
static void test_c_##OP() { \
	register float fs, ft; \
	 \
	printf("%s:\n", #OP); \
	 \
	FCMP_OP_DO_II(OP, 0x00000000, 0x00000000); \
	FCMP_OP_DO_II(OP, 0x00000000, 0x80000000); \
	FCMP_OP_DO_II(OP, 0x80000000, 0x00000000); \
	FCMP_OP_DO_II(OP, 0x80000000, 0x80000000); \
	FCMP_OP_DO_II(OP, 0x00000000, 0x3F800000); \
	FCMP_OP_DO_II(OP, 0x3F800000, 0x3F800000); \
	FCMP_OP_DO_II(OP, 0x3F800000, 0x00000000); \
	FCMP_OP_DO_II(OP, 0x40000000, 0x40000000); \
	FCMP_OP_DO_II(OP, 0x40400000, 0x3F800000); \
	FCMP_OP_DO_II(OP, 0x40400000, 0x40400000); \
	FCMP_OP_DO_II(OP, 0x7FFFFFFF, 0x7FFFFFFF); \
	FCMP_OP_DO_II(OP, 0x7FFFFFFF, 0xFFFFFFFF); \
	FCMP_OP_DO_II(OP, 0xFFFFFFFF, 0x7FFFFFFF); \
	FCMP_OP_DO_II(OP, 0xFFFFFFFF, 0xFFFFFFFF); \
	FCMP_OP_DO_II(OP, 0x7FFFFFFF, 0x00000000); \
	FCMP_OP_DO_II(OP, 0x00000000, 0x00000001); \
	FCMP_OP_DO_II(OP, 0x00000001, 0x00000000); \
	FCMP_OP_DO_II(OP, 0x00000000, 0x7FFFFFFF); \
	FCMP_OP_DO_II(OP, 0x00000000, 0x7F800000); \
	FCMP_OP_DO_II(OP, 0x7F800000, 0x7F800000); \
	FCMP_OP_DO_II(OP, 0x7F800000, 0x7F800001); \
	FCMP_OP_DO_II(OP, 0x7F800001, 0x7F800000); \
	FCMP_OP_DO_II(OP, 0xFF800000, 0x7F800000); \
	 \
	FCMP_OP_DO_MM(OP, CF_ZERO, CF_ZERO); \
	FCMP_OP_DO_MM(OP, CF_ZERO, CF_NEGZERO); \
	FCMP_OP_DO_MM(OP, CF_ONE, CF_ONE); \
	FCMP_OP_DO_MM(OP, CF_ONE, CF_NEGONE); \
	FCMP_OP_DO_MM(OP, CF_MAX_MANTISSA, CF_ONE); \
	FCMP_OP_DO_MM(OP, CF_MAX_EXP, CF_ONE); \
	FCMP_OP_DO_MM(OP, CF_MIN_EXP, CF_ONE); \
	FCMP_OP_DO_MM(OP, CF_ONE, CF_MAX_MANTISSA); \
	FCMP_OP_DO_MM(OP, CF_ONE, CF_MAX_EXP); \
	FCMP_OP_DO_MM(OP, CF_ONE, CF_MIN_EXP); \
	FCMP_OP_DO_MM(OP, CF_MAX, CF_ONE); \
	FCMP_OP_DO_MM(OP, CF_MIN, CF_ONE); \
	FCMP_OP_DO_MM(OP, CF_ONE, CF_MAX); \
	FCMP_OP_DO_MM(OP, CF_ONE, CF_MIN); \
	FCMP_OP_DO_MM(OP, CF_MAX, CF_MAX); \
	FCMP_OP_DO_MM(OP, CF_MIN, CF_MIN); \
	FCMP_OP_DO_MM(OP, CF_GARBAGE1, CF_GARBAGE2); \
	\
	printf("\n"); \
}

TEST_FCMP_FUNC(eq)
TEST_FCMP_FUNC(f)
TEST_FCMP_FUNC(le)
TEST_FCMP_FUNC(lt)

int main(int argc, char **argv) {
	printf("-- TEST BEGIN\n");

	test_c_eq();
	test_c_f();
	test_c_le();
	test_c_lt();

	printf("-- TEST END\n");

	return 0;
}
