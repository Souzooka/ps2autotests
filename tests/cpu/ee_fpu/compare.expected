-- TEST BEGIN
eq:
  eq 00000000/+0.00, 00000000/+0.00: true
  eq 00000000/+0.00, 80000000/-0.00: true
  eq 80000000/-0.00, 00000000/+0.00: true
  eq 80000000/-0.00, 80000000/-0.00: true
  eq 00000000/+0.00, 3f800000/+1.00: false
  eq 3f800000/+1.00, 3f800000/+1.00: true
  eq 3f800000/+1.00, 00000000/+0.00: false
  eq 40000000/+2.00, 40000000/+2.00: true
  eq 40400000/+3.00, 3f800000/+1.00: false
  eq 40400000/+3.00, 40400000/+3.00: true
  eq 7fffffff/+NaN, 7fffffff/+NaN: true
  eq 7fffffff/+NaN, ffffffff/-NaN: false
  eq ffffffff/-NaN, 7fffffff/+NaN: false
  eq ffffffff/-NaN, ffffffff/-NaN: true
  eq 7fffffff/+NaN, 00000000/+0.00: false
  eq 00000000/+0.00, 00000001/+0.00: true
  eq 00000001/+0.00, 00000000/+0.00: true
  eq 00000000/+0.00, 7fffffff/+NaN: false
  eq 00000000/+0.00, 7f800000/+NaN: false
  eq 7f800000/+NaN, 7f800000/+NaN: true
  eq 7f800000/+NaN, 7f800001/+NaN: false
  eq 7f800001/+NaN, 7f800000/+NaN: false
  eq ff800000/-NaN, 7f800000/+NaN: false
  eq CF_ZERO, CF_ZERO: true
  eq CF_ZERO, CF_NEGZERO: true
  eq CF_ONE, CF_ONE: true
  eq CF_ONE, CF_NEGONE: false
  eq CF_MAX_MANTISSA, CF_ONE: false
  eq CF_MAX_EXP, CF_ONE: false
  eq CF_MIN_EXP, CF_ONE: false
  eq CF_ONE, CF_MAX_MANTISSA: false
  eq CF_ONE, CF_MAX_EXP: false
  eq CF_ONE, CF_MIN_EXP: false
  eq CF_MAX, CF_ONE: false
  eq CF_MIN, CF_ONE: false
  eq CF_ONE, CF_MAX: false
  eq CF_ONE, CF_MIN: false
  eq CF_MAX, CF_MAX: true
  eq CF_MIN, CF_MIN: true
  eq CF_GARBAGE1, CF_GARBAGE2: false

f:
  f 00000000/+0.00, 00000000/+0.00: false
  f 00000000/+0.00, 80000000/-0.00: false
  f 80000000/-0.00, 00000000/+0.00: false
  f 80000000/-0.00, 80000000/-0.00: false
  f 00000000/+0.00, 3f800000/+1.00: false
  f 3f800000/+1.00, 3f800000/+1.00: false
  f 3f800000/+1.00, 00000000/+0.00: false
  f 40000000/+2.00, 40000000/+2.00: false
  f 40400000/+3.00, 3f800000/+1.00: false
  f 40400000/+3.00, 40400000/+3.00: false
  f 7fffffff/+NaN, 7fffffff/+NaN: false
  f 7fffffff/+NaN, ffffffff/-NaN: false
  f ffffffff/-NaN, 7fffffff/+NaN: false
  f ffffffff/-NaN, ffffffff/-NaN: false
  f 7fffffff/+NaN, 00000000/+0.00: false
  f 00000000/+0.00, 00000001/+0.00: false
  f 00000001/+0.00, 00000000/+0.00: false
  f 00000000/+0.00, 7fffffff/+NaN: false
  f 00000000/+0.00, 7f800000/+NaN: false
  f 7f800000/+NaN, 7f800000/+NaN: false
  f 7f800000/+NaN, 7f800001/+NaN: false
  f 7f800001/+NaN, 7f800000/+NaN: false
  f ff800000/-NaN, 7f800000/+NaN: false
  f CF_ZERO, CF_ZERO: false
  f CF_ZERO, CF_NEGZERO: false
  f CF_ONE, CF_ONE: false
  f CF_ONE, CF_NEGONE: false
  f CF_MAX_MANTISSA, CF_ONE: false
  f CF_MAX_EXP, CF_ONE: false
  f CF_MIN_EXP, CF_ONE: false
  f CF_ONE, CF_MAX_MANTISSA: false
  f CF_ONE, CF_MAX_EXP: false
  f CF_ONE, CF_MIN_EXP: false
  f CF_MAX, CF_ONE: false
  f CF_MIN, CF_ONE: false
  f CF_ONE, CF_MAX: false
  f CF_ONE, CF_MIN: false
  f CF_MAX, CF_MAX: false
  f CF_MIN, CF_MIN: false
  f CF_GARBAGE1, CF_GARBAGE2: false

le:
  le 00000000/+0.00, 00000000/+0.00: true
  le 00000000/+0.00, 80000000/-0.00: true
  le 80000000/-0.00, 00000000/+0.00: true
  le 80000000/-0.00, 80000000/-0.00: true
  le 00000000/+0.00, 3f800000/+1.00: true
  le 3f800000/+1.00, 3f800000/+1.00: true
  le 3f800000/+1.00, 00000000/+0.00: false
  le 40000000/+2.00, 40000000/+2.00: true
  le 40400000/+3.00, 3f800000/+1.00: false
  le 40400000/+3.00, 40400000/+3.00: true
  le 7fffffff/+NaN, 7fffffff/+NaN: true
  le 7fffffff/+NaN, ffffffff/-NaN: false
  le ffffffff/-NaN, 7fffffff/+NaN: true
  le ffffffff/-NaN, ffffffff/-NaN: true
  le 7fffffff/+NaN, 00000000/+0.00: false
  le 00000000/+0.00, 00000001/+0.00: true
  le 00000001/+0.00, 00000000/+0.00: true
  le 00000000/+0.00, 7fffffff/+NaN: true
  le 00000000/+0.00, 7f800000/+NaN: true
  le 7f800000/+NaN, 7f800000/+NaN: true
  le 7f800000/+NaN, 7f800001/+NaN: true
  le 7f800001/+NaN, 7f800000/+NaN: false
  le ff800000/-NaN, 7f800000/+NaN: true
  le CF_ZERO, CF_ZERO: true
  le CF_ZERO, CF_NEGZERO: true
  le CF_ONE, CF_ONE: true
  le CF_ONE, CF_NEGONE: false
  le CF_MAX_MANTISSA, CF_ONE: false
  le CF_MAX_EXP, CF_ONE: false
  le CF_MIN_EXP, CF_ONE: true
  le CF_ONE, CF_MAX_MANTISSA: true
  le CF_ONE, CF_MAX_EXP: true
  le CF_ONE, CF_MIN_EXP: false
  le CF_MAX, CF_ONE: false
  le CF_MIN, CF_ONE: true
  le CF_ONE, CF_MAX: true
  le CF_ONE, CF_MIN: false
  le CF_MAX, CF_MAX: true
  le CF_MIN, CF_MIN: true
  le CF_GARBAGE1, CF_GARBAGE2: false

lt:
  lt 00000000/+0.00, 00000000/+0.00: false
  lt 00000000/+0.00, 80000000/-0.00: false
  lt 80000000/-0.00, 00000000/+0.00: false
  lt 80000000/-0.00, 80000000/-0.00: false
  lt 00000000/+0.00, 3f800000/+1.00: true
  lt 3f800000/+1.00, 3f800000/+1.00: false
  lt 3f800000/+1.00, 00000000/+0.00: false
  lt 40000000/+2.00, 40000000/+2.00: false
  lt 40400000/+3.00, 3f800000/+1.00: false
  lt 40400000/+3.00, 40400000/+3.00: false
  lt 7fffffff/+NaN, 7fffffff/+NaN: false
  lt 7fffffff/+NaN, ffffffff/-NaN: false
  lt ffffffff/-NaN, 7fffffff/+NaN: true
  lt ffffffff/-NaN, ffffffff/-NaN: false
  lt 7fffffff/+NaN, 00000000/+0.00: false
  lt 00000000/+0.00, 00000001/+0.00: false
  lt 00000001/+0.00, 00000000/+0.00: false
  lt 00000000/+0.00, 7fffffff/+NaN: true
  lt 00000000/+0.00, 7f800000/+NaN: true
  lt 7f800000/+NaN, 7f800000/+NaN: false
  lt 7f800000/+NaN, 7f800001/+NaN: true
  lt 7f800001/+NaN, 7f800000/+NaN: false
  lt ff800000/-NaN, 7f800000/+NaN: true
  lt CF_ZERO, CF_ZERO: false
  lt CF_ZERO, CF_NEGZERO: false
  lt CF_ONE, CF_ONE: false
  lt CF_ONE, CF_NEGONE: false
  lt CF_MAX_MANTISSA, CF_ONE: false
  lt CF_MAX_EXP, CF_ONE: false
  lt CF_MIN_EXP, CF_ONE: true
  lt CF_ONE, CF_MAX_MANTISSA: true
  lt CF_ONE, CF_MAX_EXP: true
  lt CF_ONE, CF_MIN_EXP: false
  lt CF_MAX, CF_ONE: false
  lt CF_MIN, CF_ONE: true
  lt CF_ONE, CF_MAX: true
  lt CF_ONE, CF_MIN: false
  lt CF_MAX, CF_MAX: false
  lt CF_MIN, CF_MIN: false
  lt CF_GARBAGE1, CF_GARBAGE2: false

-- TEST END
